src/coreclr/gc/gcpriv.h

```diff
@@ -4299,142 +4489,472 @@ class gc_heap
         {
-            total_recorded_tcp = 0;
-            recorded_tcp_index = 0;
-            dprintf (6666, ("INIT tcp buffer"));
-        }
+            // If this is right after we adjusted, we should see if we were successful with the adjustment.
+            size_t last_changed_gc_index = adj->gc_index;
+            if (!last_changed_gc_index) return;
 
-        int get_recorded_tcp_count () { return total_recorded_tcp; }
+            bool check_p = (current_gc_index < (last_changed_gc_index + (2 * sample_size)));
 
-        //
-        // Maintain some info about last time we did change heap count.
-        //
-        size_t          last_changed_gc_index;
-        // This is intentionally kept as a float for precision.
-        float           last_changed_count;
-        float           last_changed_stcp;
+            dprintf (6666, ("last adjusted at GC#%Id, %Id GCs ago, %s",
+                last_changed_gc_index, (current_gc_index - last_changed_gc_index), (check_p ? "check success" : "already checked success")));
+            if (!check_p)
+            {
+                return;
+            }
 
-        //
-        // For tuning above/below target tcp.
-        //
-        // If we just increased the heap count and immediately need to grow again, that counts as a failure.
-        // The higher the failure count, the more aggressive we should grow.
-        int             inc_failure_count;
-
-        // If we are trending up and the tcp is already close enough to target, we need this many samples
-        // before we adjust.
-        int             inc_recheck_threshold;
-
-        // If we shrink and the stcp doesn't change much, that counts as a failure. For the below target case
-        // it's fine to stay here for a while. Either it'll naturally change and break out of this situation
-        // or we wait for a while before we re-evaluate. How long we wait is defined by dec_recheck_threshold
-        // each time our calculation tells us to shrink.
-        int             dec_failure_count;
-        int             dec_failure_recheck_threshold;
-
-        // If we continue to be below target for an extended period of time, ie, we've accumulated more than
-        // below_target_threshold, we want to reduce the heap count.
-        float           below_target_accumulation;
-        float           below_target_threshold;
-
-        // TODO: we should refactor this and the inc checks into a utility class.
-        bool            dec_by_one_scheduled;
-        int             dec_by_one_count;
-
-        // Currently only used for dprintf.
-        size_t          first_below_target_gc_index;
-
-        float get_range_upper (float t)
-        {
-            return (t * 1.2f);
+            adjust_metric adj_metric = adj->metric;
+
+            // We are guaranteed to have at least sample_size amount of new samples.
+            if (adj_metric == adjust_hc)
```

Comment:
does this need an assert to ensure there are at least sample_size new samples?

Reply:
I could add an assert. probably in a later PR if we want to merge this today.

Reply:
ok sounds good.